docs/proposal/add_support_for_cache.md

```diff
@@ -0,0 +1,360 @@
+3. **Fetch Repository**:
+   - Cargo checks if the repository is cached in `~/.cargo/git/db/`.
+   - If not cached, Cargo clones the repository as a bare repository:
+     ```plaintext
+     ~/.cargo/git/db/
+     ├── my-crate-<hash>/
+     │   ├── objects/
+     │   ├── refs/
+     │   └── ... (bare repository files)
+     ```
+
+4. **Create Checkout**:
+   - Cargo creates a working copy in `~/.cargo/git/checkouts/` for the specific commit:
+     ```plaintext
+     ~/.cargo/git/checkouts/
+     ├── my-crate-<hash>/
+     │   ├── src/
+     │   ├── .git/
+     │   └── ... (repository files at commit a1b2c3d4)
+     ```
+
+5. **Build the Dependency**:
+   - Cargo uses the working copy to build the project with the specified commit.
+
+6. **Reuse and Update**:
+   - For subsequent builds, Cargo reuses the cached bare repository.
+   - If the dependency is updated to a new commit, Cargo fetches updates to the bare repository and creates a new checkout:
+     ```toml
+     [dependencies]
+     my-crate = { git = "https://github.com/user/my-crate.git", rev = "e5f6g7h8" }
+     ```
+     ```sh
+     cargo build
+     ```
+
+By adopting a similar approach, KPM can provide KCL developers with a unified and efficient system for managing third-party dependencies, abstracting the complexities of downloading, caching, and version control.
+
+## 3. Proposed Changes to KPM
+
+### New Directory Structure
+```bash
+└── kpm
+    ├── .kpm # all the configuration file for kpm client
+    ├── git # all the kcl dependencies from git repo
+    ├── oci # all the kcl dependencies from oci registry
+```
+Under the subdir in `kpm/git`, the storage local system is
+```bash
+kpm/git
+├── checkouts.   # checkout the specific version of git repository from cache bare repository 
+│   ├── kcl-2a81898195a215f1
+│   │   └── 33bb450. . # All the version of kcl package from git repository will be replaced with commit id
+│   ├── kcl-578669463c900b87
+│   │   └── 33bb450
+└── db    # A bare git repository for cache git repo
+    ├── kcl-2a81898195a215f1.      # <NAME>-<HASH> <NAME> is the name of git repo, 
+    ├── kcl-578669463c900b87.   # <HASH> is calculated by the git full url.
+```
+Under the subdir in `kpm/oci`, the storage local system is
+```bash
+kpm/oci
+├── cache # the cache for KCL dependencies tar
+│   ├── ghcr.io-2a81898195a215f1    # <HOST>-<HASH> HOST is the name of oci registry,  <HASH> is calculated by the oci full url.
+│   │   └── k8s_1.29.tar    # the tar for KCL dependencies
+│   ├── docker.io-578669463c900b87
+│   │   └── k8s_1.28.tar
+└── src                                              	
+│   ├── ghcr.io-2a81898195a215f1
+│   │   └── k8s_1.29    # the KCL dependencies tar will untar here
+│   ├── docker.io-578669463c900b87
+│   │   └── k8s_1.28
+```
+### 1. [PRETEST 1]: Enhance git module to support bare repository clone
+
+**Completed** ✅ - [PRETEST]: [Added support for bare repo in clone function of git module](https://github.com/kcl-lang/kpm/pull/419)
```

Comment:
Do not delete the previous content in this part, you can rewrite this part of the content to align with the PR, but do not delete.